main.cpp

```diff
@@ -253,481 +202,222 @@ Use a service like https://www.diffchecker.com/diff to compare your output.
 
-struct Point
-{
-    ~Point();
-
-    Point(const FloatType& _x, const FloatType& _y);
-    Point(const DoubleType& _x, const DoubleType& _y);
-    Point(const IntType& _x, const IntType& _y);
-
-    Point& multiply(float m);
-    Point& multiply(FloatType& f);
-    Point& multiply(DoubleType& d);
-    Point& multiply(IntType& i);
-
-    void toString();
-
-private:
-    float x{0}, y{0};
-};
-
-struct A {};
-struct HeapA
-{
-    A* pointerToA = nullptr;
+    Numeric(Type t) : value( std::make_unique<Type>(t) ) { }
 
-    HeapA() : pointerToA(new A) {}
-
-    ~HeapA()
+    ~Numeric() 
     {
-        delete pointerToA;
+        value = nullptr;
     }
 
-};
-
-struct FloatType;
-struct DoubleType;
-struct IntType;
-
-struct FloatType
-{
-    FloatType(float floatPrimitive);
-    ~FloatType();
-
-    FloatType& operator+=(float f);
-    FloatType& operator-=(float f);
-    FloatType& operator*=(float f);
-    FloatType& operator/=(float f);
-
-    operator float() const { return *value; }
-
-    FloatType& pow(float f);
-    FloatType& pow(const IntType& i);
-    FloatType& pow(const FloatType& f);
-    FloatType& pow(const DoubleType& d);
-
-    /*
-        1) add two member functions named "apply()" to each of your Heap-Allocated Numeric Type wrappers.
-    */
+    operator Type() const { return *value; }
 
-    //  both apply() functions should work with chaining
-    //  2) One of the apply() functions should takes a std::function<> object as the function argument.
-    //  the std::function<> object should return *this;
-    FloatType& apply(std::function<FloatType&(float&)> func);
-
-    //  3) the other apply() function should take a function pointer. 
-    //  the function pointer should return void.
-    FloatType& apply(void(*func)(float&));
-
-private:
-    float* value = nullptr;
-    FloatType& powInternal(const float f);
-};
-
-struct DoubleType
-{
-    DoubleType(double doublePrimitive);
-    ~DoubleType();
-
-    DoubleType& operator+=(double d);
-    DoubleType& operator-=(double d);
-    DoubleType& operator*=(double d);
-    DoubleType& operator/=(double d);
+    Numeric& operator+=(Type t)
+    {
+        *value += t;
+        return *this;
+    }
 
-    operator double() const { return *value; }
+    Numeric& operator-=(Type t)
+    {
+        *value -= t;
+        return *this;
+    }
 
-    DoubleType& pow(double d);
-    DoubleType& pow(const IntType& i);
-    DoubleType& pow(const FloatType& f);
-    DoubleType& pow(const DoubleType& d);
+    Numeric& operator*=(Type t)
+    {
+        *value *= t;
+        return *this;
+    }
 
-    DoubleType& apply(std::function<DoubleType&(double&)> func);
-    DoubleType& apply(void(*func)(double&));
+    // #11
+    Numeric& operator/=(Type t)
+    {
+        // template type is an int
+        if constexpr (std::is_same<int, Type>::value)
+        {
+            // parameter's type is also an int
+            if constexpr (std::is_same<int, decltype(t)>::value)
```

Comment:
Your parameter's type can be something other than int.  that means you need to template this function.
- [ ] revise

Reply:
The struct template typename is NumericType. I am going with my assumption that instruction  #11 meant to compare that templated type to type int in operator/=  

That was actually my initial idea and changed it after looking at 3 other students examples which compare the operand parameter type to int but not the template type.

like this here compares the _function parameter's type_ not the _struct template's type_:   

std::is_same<Type, int>

So now I am now doing following instead:

std::is_same<NumericType, int>

I saw the following from 3 other students examples uses decltype() to check the _declared_ type of t:

if constexpr (std::is_same<int, decltype(t)>::value)

However, in my code that causes a Floating point exception (core dumped)

Also the instruction #11 states to compare the parameter's type which I am now doing like this:

if constexpr (std::is_same<int, Type>::value)